File: maxon-zig/src/ast.zig
const std = @import("std");

pub const Program = struct {
    functions: []FunctionDecl,
};

pub const FunctionDecl = struct {
    name: []const u8,
    return_type: ?[]const u8, // null for void
    body: []Statement,
};

pub const Statement = union(enum) {
    @"return": ReturnStmt,
};

pub const ReturnStmt = struct {
    value: ?Expression,
};

pub const Expression = union(enum) {
    integer: i64,
};

File: maxon-sharp/Compiler/Parser/Ast.cs
namespace MaxonSharp.Parser;

// Type references
public abstract record TypeRef;
public record IntTypeRef : TypeRef;

// Expressions
public abstract record Expr;
public record IntLiteralExpr(long Value) : Expr;
public record IdentifierExpr(string Name) : Expr;

// Statements
public abstract record Stmt;
public record ReturnStmt(Expr? Value) : Stmt;

// Declarations
public record FunctionDecl(string Name, TypeRef? ReturnType, List<Stmt> Body);

// Program (root)
public record ProgramAst(List<FunctionDecl> Functions);
